// @ui5/webcomponents-react/lib/DateRangePicker.js
import { CalendarType } from '@ui5/webcomponents-react/lib/CalendarType';
import { ValueState } from '@ui5/webcomponents-react/lib/ValueState';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/DateRangePicker';

/**
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/DateRangePicker" target="_blank">UI5 Web Components Playground</a>
 */
var DateRangePicker = withWebComponent('ui5-daterange-picker', ['delimiter', 'formatPattern', 'maxDate', 'minDate', 'name', 'placeholder', 'primaryCalendarType', 'value', 'valueState'], ['disabled', 'hideWeekNumbers', 'readonly'], ['valueStateMessage'], ['change', 'input']);
DateRangePicker.displayName = 'DateRangePicker';
DateRangePicker.defaultProps = {
  delimiter: '-',
  disabled: false,
  hideWeekNumbers: false,
  placeholder: undefined,
  primaryCalendarType: CalendarType.Gregorian,
  readonly: false,
  valueState: ValueState.None
};

export { DateRangePicker };
