// @ui5/webcomponents-react/lib/ToolbarSeparator.js
import { createComponentStyles } from '@ui5/webcomponents-react-base/lib/createComponentStyles';
import { CssSizeVariables } from '@ui5/webcomponents-react-base/lib/CssSizeVariables';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';
import React from 'react';

var styles = {
  separator: {
    width: '0.0625rem',
    height: CssSizeVariables.sapWcrToolbarSeparatorHeight,
    background: ThemingParameters.sapToolbar_SeparatorColor
  }
};
var useStyles = createComponentStyles(styles, {
  name: 'ToolbarSeparator'
});

var ToolbarSeparator = function ToolbarSeparator(props) {
  var style = props.style,
      className = props.className;
  var classes = useStyles(styles);
  var separatorClasses = StyleClassHelper.of(classes.separator);

  if (className) {
    separatorClasses.put(className);
  }

  return /*#__PURE__*/React.createElement("div", {
    style: style,
    className: separatorClasses.valueOf()
  });
};

ToolbarSeparator.displayName = 'ToolbarSeparator';

export { ToolbarSeparator };
