// hgi-logger/index.js
const path = require('path');
const debug = require('hgi-debug')

function isJSON(data) {
  if(typeof data !== 'string')
    return false;
  try {
    const result = JSON.parse(data);
    const type = Object.prototype.toString.call(result);
    return type === '[object Object]' || 
           type === '[object Array]'
  } catch(e) {
    return false;
  }
}

module.exports.error = async function error(data) {

  let message = {};
  let jsonFlag = isJSON(data)

  if(!jsonFlag) {
    message.containerId = path.basename(process.cwd());
    message.status = 'ERROR';
    message.data = data;
  } else {
    message = data;
  }

  debug.log('hgi-logger: catched error - ', data)

  fetch('http://logger', {
    method: 'POST',
    mode: 'no-cors',
    cache: 'no-cache',
    credentials: 'omit',
    headers: {
      'Content-Type': 'application/json'
    },
    redirect: 'error',
    referrerPolicy: 'no-referrer',
    body: JSON.stringify(message)
  }).catch(error => {
    if(jsonFlag)
      console.log(JSON.stringify(data))
    else
      console.log(data);
  });  
}
