// @ui5/webcomponents-react/lib/AnalyticalCard.js
import _extends from '@babel/runtime/helpers/extends';
import { createUseStyles } from 'react-jss';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import React, { forwardRef } from 'react';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';
import { sapUiContentPadding } from '@ui5/webcomponents-react-base/lib/spacing';

var styles = {
  card: {
    backgroundColor: ThemingParameters.sapTile_Background,
    border: "0.0625rem solid ".concat(ThemingParameters.sapTile_BorderColor),
    boxShadow: ThemingParameters.sapContent_Shadow0,
    borderRadius: ThemingParameters.sapElement_BorderCornerRadius,
    textAlign: 'start',
    overflow: 'hidden',
    position: 'relative',
    fontFamily: ThemingParameters.sapFontFamily,
    boxSizing: 'border-box',
    width: '20rem'
  },
  content: sapUiContentPadding
};

var useStyles = createUseStyles(styles, {
  name: 'AnalyticalCard'
});
/**
 * The `AnalyticalCard` is mainly used for data visualization. It consists of two areas – a header area and a chart area with a visual representation of the data.<br />
 */

var AnalyticalCard = /*#__PURE__*/forwardRef(function (props, ref) {
  var children = props.children,
      style = props.style,
      className = props.className,
      tooltip = props.tooltip,
      header = props.header;
  var classes = useStyles();
  var classNameString = StyleClassHelper.of(classes.card);

  if (className) {
    classNameString.put(className);
  }

  var passThroughProps = usePassThroughHtmlProps(props);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    className: classNameString.toString(),
    style: style,
    title: tooltip
  }, passThroughProps), header, /*#__PURE__*/React.createElement("div", {
    className: classes.content
  }, children));
});
AnalyticalCard.displayName = 'AnalyticalCard';

export { AnalyticalCard };
