// @ui5/webcomponents-react/lib/Breadcrumbs.js
import _extends from '@babel/runtime/helpers/extends';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { BreadcrumbsSeparatorStyle } from '@ui5/webcomponents-react/lib/BreadcrumbsSeparatorStyle';
import { Label } from '@ui5/webcomponents-react/lib/Label';
import React, { forwardRef, Children, Fragment } from 'react';

var SeparatorStyles = {
  BackSlash: '\\',
  DoubleBackSlash: '\\\\',
  DoubleGreaterThan: '>>',
  DoubleSlash: '//',
  GreaterThan: '>',
  Slash: '/'
};
var separatorInlineStyles = {
  margin: '0 0.25rem'
};

/**
 * Enables users to navigate between items by providing a list of links to previous steps in the user's navigation path.
 */
var Breadcrumbs = /*#__PURE__*/forwardRef(function (props, ref) {
  var children = props.children,
      separatorStyle = props.separatorStyle,
      currentLocationText = props.currentLocationText,
      tooltip = props.tooltip,
      style = props.style,
      className = props.className,
      slot = props.slot;
  var childrenArray = Children.toArray(children).filter(Boolean);
  var passThroughProps = usePassThroughHtmlProps(props);
  return /*#__PURE__*/React.createElement("div", _extends({
    ref: ref,
    title: tooltip,
    style: style,
    className: className,
    slot: slot
  }, passThroughProps), childrenArray.map(function (item, index) {
    if (index === childrenArray.length - 1) {
      return item;
    }

    return /*#__PURE__*/React.createElement(Fragment, {
      key: index
    }, item, /*#__PURE__*/React.createElement(Label, {
      style: separatorInlineStyles,
      children: SeparatorStyles[separatorStyle]
    }));
  }), currentLocationText && /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Label, {
    style: separatorInlineStyles
  }, SeparatorStyles[separatorStyle]), /*#__PURE__*/React.createElement(Label, null, currentLocationText)));
});
Breadcrumbs.displayName = 'Breadcrumbs';
Breadcrumbs.defaultProps = {
  separatorStyle: BreadcrumbsSeparatorStyle.Slash
};

export { Breadcrumbs };
