// @ui5/webcomponents-react/lib/ComboBox.js
import { ValueState } from '@ui5/webcomponents-react/lib/ValueState';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/ComboBox';

/**
 * The `ComboBox` component represents a drop-down menu with a list of the available options and a text input field to narrow down the options. It is commonly used to enable users to select one or more options from a predefined list.
 *
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/ComboBox" target="_blank">UI5 Web Components Playground</a>
 */
var ComboBox = withWebComponent('ui5-combobox', ['filter', 'filterValue', 'placeholder', 'value', 'valueState'], ['disabled', 'loading', 'readonly', 'required'], ['icon', 'valueStateMessage'], ['change', 'input']);
ComboBox.displayName = 'ComboBox';
ComboBox.defaultProps = {
  disabled: false,
  filter: 'StartsWithPerTerm',
  loading: false,
  readonly: false,
  required: false,
  valueState: ValueState.None
};

export { ComboBox };
