// @ui5/webcomponents-react/lib/DurationPicker.js
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/DurationPicker';

/**
 * The `DurationPicker` component provides an input field with assigned sliders which opens on user action. The `DurationPicker` allows users to select a time duration. It consists of two parts: the time input field and the sliders.
 *
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/DurationPicker" target="_blank">UI5 Web Components Playground</a>
 */
var DurationPicker = withWebComponent('ui5-duration-picker', ['maxValue', 'minutesStep', 'secondsStep', 'value', 'valueState'], ['disabled', 'hideHours', 'hideMinutes', 'hideSeconds', 'readonly'], ['valueStateMessage'], ['change']);
DurationPicker.displayName = 'DurationPicker';
DurationPicker.defaultProps = {
  disabled: false,
  hideHours: false,
  hideMinutes: false,
  hideSeconds: false,
  maxValue: '23:59:59',
  minutesStep: 1,
  readonly: false,
  secondsStep: 1,
  value: '00:00:00',
  valueState: 'None'
};

export { DurationPicker };
