// @ui5/webcomponents-react/lib/DynamicPageHeader.js
import _extends from '@babel/runtime/helpers/extends';
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import { createUseStyles } from 'react-jss';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import { FlexBox } from '@ui5/webcomponents-react/lib/FlexBox';
import { FlexBoxAlignItems } from '@ui5/webcomponents-react/lib/FlexBoxAlignItems';
import React, { forwardRef, useMemo } from 'react';
import { sapUiResponsiveContentPadding } from '@ui5/webcomponents-react-base/lib/spacing';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { isIE } from '@ui5/webcomponents-react-base/lib/Device';

var _headerCollapsed;
var DynamicPageCssVariables = {
  headerDisplay: '--ui5wcr_DynamicPage_header_display'
};
({
  dynamicPage: _defineProperty({
    width: '100%',
    height: '100%',
    maxHeight: '100%',
    position: 'relative',
    whiteSpace: 'normal',
    fontFamily: ThemingParameters.sapFontFamily,
    backgroundColor: ThemingParameters.sapBackgroundColor,
    overflowX: 'hidden',
    overflowY: 'auto'
  }, DynamicPageCssVariables.headerDisplay, 'block'),
  headerCollapsed: (_headerCollapsed = {}, _defineProperty(_headerCollapsed, DynamicPageCssVariables.headerDisplay, 'none'), _defineProperty(_headerCollapsed, '& $anchorBar > section[role="navigation"]', {
    top: '-0.025rem'
  }), _headerCollapsed),
  anchorBar: {
    position: 'sticky',
    boxShadow: ThemingParameters.sapContent_HeaderShadow,
    height: '1px',
    zIndex: 2,
    '& > section[role="navigation"]': {
      top: 0
    }
  },
  iEClass: {
    position: 'fixed',
    //page width - scrollbar
    width: 'calc(100% - 18px)'
  },
  iEBackgroundElement: {
    position: 'fixed',
    backgroundColor: ThemingParameters.sapObjectHeader_Background
  },
  contentContainer: _objectSpread(_objectSpread({}, sapUiResponsiveContentPadding), {}, {
    paddingTop: '1rem !important',
    boxSizing: 'border-box',
    width: '100%',
    height: 'auto',
    fontFamily: ThemingParameters.sapFontFamily,
    zIndex: 0
  }),
  backgroundStandard: {
    background: ThemingParameters.sapBackgroundColor
  },
  backgroundSolid: {
    background: ThemingParameters.sapBackgroundColor
  },
  backgroundList: {
    background: ThemingParameters.sapGroup_ContentBackground
  },
  backgroundTransparent: {
    background: 'transparent'
  }
});

var padding = isIE() ? {
  padding: '0 2rem'
} : _objectSpread({}, sapUiResponsiveContentPadding);
var DynamicPageHeaderStyles = {
  header: _objectSpread(_objectSpread({}, padding), {}, {
    flexShrink: 0,
    backgroundColor: ThemingParameters.sapObjectHeader_Background,
    position: 'sticky',
    zIndex: 1,
    paddingTop: '1rem !important',
    paddingBottom: '1rem !important',
    display: "var(".concat(DynamicPageCssVariables.headerDisplay, ")"),
    maxHeight: '500px',
    overflow: isIE() ? 'visible' : 'hidden'
  }),
  iEClass: {
    position: 'fixed',
    // page width - padding - scrollbar
    width: 'calc(100% - 18px - 4rem)'
  }
};

var useStyles = createUseStyles(DynamicPageHeaderStyles, {
  name: 'DynamicPageHeader'
});
/**
 * The dynamic page header contains the header content of the dynamic page.
 * This component can be collapsed and pinned by the anchorbar.
 */

var DynamicPageHeader = /*#__PURE__*/forwardRef(function (props, ref) {
  var children = props.children,
      headerPinned = props.headerPinned,
      topHeaderHeight = props.topHeaderHeight,
      tooltip = props.tooltip,
      className = props.className,
      style = props.style;
  var passThroughProps = usePassThroughHtmlProps(props);
  var headerStyles = useMemo(function () {
    if (headerPinned) {
      return _objectSpread(_objectSpread({}, style), {}, {
        top: "".concat(topHeaderHeight, "px"),
        zIndex: 1
      });
    }

    return style;
  }, [headerPinned, topHeaderHeight, style]);
  var classes = useStyles();
  var classNames = StyleClassHelper.of(classes.header);

  if (isIE()) {
    classNames.put(classes.iEClass);
  }

  classNames.putIfPresent(className);
  return /*#__PURE__*/React.createElement("div", _extends({
    title: tooltip,
    style: headerStyles,
    ref: ref,
    className: classNames.className,
    "data-component-name": "DynamicPageHeader"
  }, passThroughProps), /*#__PURE__*/React.createElement(FlexBox, {
    alignItems: FlexBoxAlignItems.Start
  }, children));
});
DynamicPageHeader.displayName = 'DynamicPageHeader';

export { DynamicPageHeader };
