// @ui5/webcomponents-react/lib/MessageBox.js
import _extends from '@babel/runtime/helpers/extends';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import '@ui5/webcomponents-icons/dist/hint';
import '@ui5/webcomponents-icons/dist/message-error';
import '@ui5/webcomponents-icons/dist/message-information';
import '@ui5/webcomponents-icons/dist/message-success';
import '@ui5/webcomponents-icons/dist/message-warning';
import '@ui5/webcomponents-icons/dist/question-mark';
import { createUseStyles } from 'react-jss';
import { useConsolidatedRef, useI18nBundle, usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/hooks';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { enrichEventWithDetails } from '@ui5/webcomponents-react-base/lib/Utils';
import { ABORT, CANCEL, CLOSE, DELETE, IGNORE, NO, OK, RETRY, YES, HIGHLIGHT, WARNING, SUCCESS, INFORMATION, ERROR, CONFIRMATION } from '@ui5/webcomponents-react/dist/assets/i18n/i18n-defaults';
import { Button } from '@ui5/webcomponents-react/lib/Button';
import { ButtonDesign } from '@ui5/webcomponents-react/lib/ButtonDesign';
import { Dialog } from '@ui5/webcomponents-react/lib/Dialog';
import { Icon } from '@ui5/webcomponents-react/lib/Icon';
import { MessageBoxActions } from '@ui5/webcomponents-react/lib/MessageBoxActions';
import { MessageBoxTypes } from '@ui5/webcomponents-react/lib/MessageBoxTypes';
import { Text } from '@ui5/webcomponents-react/lib/Text';
import { Title } from '@ui5/webcomponents-react/lib/Title';
import { TitleLevel } from '@ui5/webcomponents-react/lib/TitleLevel';
import React, { forwardRef, useMemo, isValidElement, useCallback, useEffect } from 'react';
import { CssSizeVariables } from '@ui5/webcomponents-react-base/lib/CssSizeVariables';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';

var stopPropagation = function stopPropagation(e) {
  e.stopPropagation();
  e.stopImmediatePropagation();
};

var style = {
  messageBox: {
    '--sapContent_Shadow0': 'none'
  },
  header: {
    display: 'flex',
    height: CssSizeVariables.sapWcrBarHeight,
    textAlign: 'center',
    width: '100%',
    alignItems: 'center',
    padding: '0.25rem 1rem',
    boxSizing: 'border-box',
    borderBottom: "1px solid var(--messageBoxBorderColor)",
    color: ThemingParameters.sapContent_LabelColor,
    fontSize: '1rem',
    '& ui5-icon:first-child': {
      padding: '0 0.25rem',
      width: '1rem',
      height: '1rem'
    },
    '&[data-type="Error"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapErrorBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapErrorBorderColor,
      '--sapContent_NonInteractiveIconColor': ThemingParameters.sapNegativeElementColor
    },
    '&[data-type="Warning"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapWarningBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapWarningBorderColor,
      '--sapContent_NonInteractiveIconColor': ThemingParameters.sapCriticalElementColor
    },
    '&[data-type="Success"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapSuccessBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapSuccessBorderColor,
      '--sapContent_NonInteractiveIconColor': ThemingParameters.sapPositiveElementColor
    },
    '&[data-type="Confirm"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapNeutralBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapNeutralBorderColor,
      '--sapContent_NonInteractiveIconColor': ThemingParameters.sapNeutralElementColor
    },
    '&[data-type="Information"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapNeutralBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapNeutralBorderColor,
      '--sapContent_NonInteractiveIconColor': ThemingParameters.sapNeutralElementColor
    },
    '&[data-type="Highlight"]': {
      '--sapPageFooter_BorderColor': ThemingParameters.sapInformationBorderColor,
      '--messageBoxBorderColor': ThemingParameters.sapInformationBorderColor,
      '--sapContent_NonInteractiveIconColor': ThemingParameters.sapInformativeElementColor
    }
  },
  content: {
    padding: '1rem'
  },
  footer: {
    height: CssSizeVariables.sapWcrBarHeight,
    width: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-end',
    padding: '0 0.5rem',
    '& > *': {
      margin: '0 0.25rem'
    },
    '& > ui5-button': {
      display: 'flex',
      minWidth: '4rem'
    }
  }
};

var useStyles = createUseStyles(style, {
  name: 'MessageBox'
});
/**
 * The `MessageBox` component provides easier methods to create a `Dialog`, such as standard alerts, confirmation dialogs, or arbitrary message dialogs.
 * For convenience, it also provides an `open` prop, so it is not necessary to attach a `ref` to open the `MessageBox`.
 */

var MessageBox = /*#__PURE__*/forwardRef(function (props, ref) {
  var _actionTranslations;

  var open = props.open,
      type = props.type,
      children = props.children,
      className = props.className,
      style = props.style,
      tooltip = props.tooltip,
      slot = props.slot,
      title = props.title,
      icon = props.icon,
      actions = props.actions,
      onClose = props.onClose;
  var dialogRef = useConsolidatedRef(ref);
  var classes = useStyles();
  var iconToRender = useMemo(function () {
    if ( /*#__PURE__*/isValidElement(icon)) return icon;

    switch (type) {
      case MessageBoxTypes.CONFIRM:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "question-mark"
        });

      case MessageBoxTypes.ERROR:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "message-error"
        });

      case MessageBoxTypes.INFORMATION:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "message-information"
        });

      case MessageBoxTypes.SUCCESS:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "message-success"
        });

      case MessageBoxTypes.WARNING:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "message-warning"
        });

      case MessageBoxTypes.HIGHLIGHT:
        return /*#__PURE__*/React.createElement(Icon, {
          name: "hint"
        });

      default:
        return null;
    }
  }, [icon, type]);
  var i18nBundle = useI18nBundle('@ui5/webcomponents-react');
  var actionTranslations = (_actionTranslations = {}, _defineProperty(_actionTranslations, MessageBoxActions.ABORT, i18nBundle.getText(ABORT)), _defineProperty(_actionTranslations, MessageBoxActions.CANCEL, i18nBundle.getText(CANCEL)), _defineProperty(_actionTranslations, MessageBoxActions.CLOSE, i18nBundle.getText(CLOSE)), _defineProperty(_actionTranslations, MessageBoxActions.DELETE, i18nBundle.getText(DELETE)), _defineProperty(_actionTranslations, MessageBoxActions.IGNORE, i18nBundle.getText(IGNORE)), _defineProperty(_actionTranslations, MessageBoxActions.NO, i18nBundle.getText(NO)), _defineProperty(_actionTranslations, MessageBoxActions.OK, i18nBundle.getText(OK)), _defineProperty(_actionTranslations, MessageBoxActions.RETRY, i18nBundle.getText(RETRY)), _defineProperty(_actionTranslations, MessageBoxActions.YES, i18nBundle.getText(YES)), _actionTranslations);

  var titleToRender = function titleToRender() {
    if (title) {
      return title;
    }

    switch (type) {
      case MessageBoxTypes.CONFIRM:
        return i18nBundle.getText(CONFIRMATION);

      case MessageBoxTypes.ERROR:
        return i18nBundle.getText(ERROR);

      case MessageBoxTypes.INFORMATION:
        return i18nBundle.getText(INFORMATION);

      case MessageBoxTypes.SUCCESS:
        return i18nBundle.getText(SUCCESS);

      case MessageBoxTypes.WARNING:
        return i18nBundle.getText(WARNING);

      case MessageBoxTypes.HIGHLIGHT:
        return i18nBundle.getText(HIGHLIGHT);

      default:
        return null;
    }
  };

  var getActions = function getActions() {
    if (actions && actions.length > 0) {
      return actions;
    }

    if (type === MessageBoxTypes.CONFIRM) {
      return [MessageBoxActions.OK, MessageBoxActions.CANCEL];
    }

    if (type === MessageBoxTypes.ERROR) {
      return [MessageBoxActions.CLOSE];
    }

    return [MessageBoxActions.OK];
  };

  var handleOnClose = useCallback(function (e) {
    var action = e.target.dataset.action;
    stopPropagation(e);
    onClose(enrichEventWithDetails(e, {
      action: action
    }));
  }, [onClose]);
  useEffect(function () {
    if (dialogRef.current) {
      if (open) {
        var _dialogRef$current$op, _dialogRef$current;

        (_dialogRef$current$op = (_dialogRef$current = dialogRef.current).open) === null || _dialogRef$current$op === void 0 ? void 0 : _dialogRef$current$op.call(_dialogRef$current);
      } else {
        var _dialogRef$current$cl, _dialogRef$current2;

        (_dialogRef$current$cl = (_dialogRef$current2 = dialogRef.current).close) === null || _dialogRef$current$cl === void 0 ? void 0 : _dialogRef$current$cl.call(_dialogRef$current2);
      }
    }
  }, [dialogRef, open]);
  var passThroughProps = usePassThroughHtmlProps(props, ['onClose']);
  var messageBoxClassNames = StyleClassHelper.of(classes.messageBox).putIfPresent(className).className;
  return /*#__PURE__*/React.createElement(Dialog, _extends({
    slot: slot,
    ref: dialogRef,
    style: style,
    tooltip: tooltip,
    className: messageBoxClassNames,
    header: /*#__PURE__*/React.createElement("header", {
      className: classes.header,
      "data-type": type
    }, iconToRender, /*#__PURE__*/React.createElement(Title, {
      level: TitleLevel.H5
    }, titleToRender())),
    footer: /*#__PURE__*/React.createElement("footer", {
      className: classes.footer
    }, getActions().map(function (action, index) {
      var _actionTranslations$a;

      return /*#__PURE__*/React.createElement(Button, {
        key: action,
        design: index === 0 ? ButtonDesign.Emphasized : ButtonDesign.Transparent,
        onClick: handleOnClose,
        "data-action": action
      }, (_actionTranslations$a = actionTranslations[action]) !== null && _actionTranslations$a !== void 0 ? _actionTranslations$a : action);
    })),
    onAfterClose: open ? handleOnClose : stopPropagation
  }, passThroughProps), /*#__PURE__*/React.createElement(Text, {
    className: classes.content
  }, children));
});
MessageBox.displayName = 'MessageBox';
MessageBox.defaultProps = {
  open: false,
  title: null,
  icon: null,
  type: MessageBoxTypes.CONFIRM,
  actions: []
};

export { MessageBox };
