// @ui5/webcomponents-react/lib/MultiComboBox.js
import { ValueState } from '@ui5/webcomponents-react/lib/ValueState';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/MultiComboBox';

/**
 * The `MultiComboBox` component provides a list box with items and a text field allowing the user to either type a value directly into the control, or choose from the list of existing items. It is a drop-down list for selecting and filtering values, commonly used to enable users to select one or more options from a predefined list. The control provides an editable input field to filter the list, and a dropdown arrow of available options. The select options in the list have checkboxes that permit multi-selection. Entered values are displayed as tokens.
 *
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/MultiComboBox" target="_blank">UI5 Web Components Playground</a>
 */
var MultiComboBox = withWebComponent('ui5-multi-combobox', ['filter', 'placeholder', 'value', 'valueState'], ['allowCustomValues', 'disabled', 'open', 'readonly', 'required'], ['icon', 'valueStateMessage'], ['change', 'input', 'open-change', 'selection-change']);
MultiComboBox.displayName = 'MultiComboBox';
MultiComboBox.defaultProps = {
  allowCustomValues: false,
  disabled: false,
  filter: 'StartsWithPerTerm',
  open: false,
  readonly: false,
  required: false,
  valueState: ValueState.None
};

export { MultiComboBox };
