// @ui5/webcomponents-react/lib/MultiInput.js
import { InputType } from '@ui5/webcomponents-react/lib/InputType';
import { ValueState } from '@ui5/webcomponents-react/lib/ValueState';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/MultiInput';

/**
 * A `MultiInput` field allows the user to enter multiple values, which are displayed as `Token`. User can choose interaction for creating tokens. Fiori Guidelines say that user should create tokens when:
 *
 * *   Type a value in the input and press enter or focus out the input field (`onChange` event is fired)
 * *   Select a value from the suggestion list
 * (`onSuggestionItemSelect` event is fired)
 *
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/MultiInput" target="_blank">UI5 Web Components Playground</a>
 */
var MultiInput = withWebComponent('ui5-multi-input', ['maxlength', 'name', 'placeholder', 'type', 'value', 'valueState'], ['showValueHelpIcon', 'disabled', 'highlight', 'readonly', 'required', 'showSuggestions'], ['tokens', 'icon', 'valueStateMessage'], ['token-delete', 'value-help-trigger', 'change', 'input', 'submit', 'suggestion-item-preview', 'suggestion-item-select', 'suggestion-scroll']);
MultiInput.displayName = 'MultiInput';
MultiInput.defaultProps = {
  showValueHelpIcon: false,
  disabled: false,
  highlight: false,
  readonly: false,
  required: false,
  showSuggestions: false,
  type: InputType.Text,
  valueState: ValueState.None
};

export { MultiInput };
