// @ui5/webcomponents-react/lib/NotificationListGroupItem.js
import { Priority } from '@ui5/webcomponents-react/lib/Priority';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents-fiori/dist/NotificationListGroupItem';

/**
 * The `NotificationListGroupItem` is a special type of list item, that unlike others can group items within self, usually `NotificationListItem` items.
 * The component consists of:
 *
 * *   `Toggle` button to expand and collapse the group
 * *   `Priority` icon to display the priority of the group
 * *   `Heading` to entitle the group
 * *   Custom actions - with the use of `NotificationAction`
 * *   Items of the group
 *
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/NotificationListGroupItem" target="_blank">UI5 Web Components Playground</a>
 */
var NotificationListGroupItem = withWebComponent('ui5-li-notification-group', ['heading', 'priority'], ['collapsed', 'showCounter', 'busy', 'showClose', 'selected'], ['actions'], ['toggle', 'close']);
NotificationListGroupItem.displayName = 'NotificationListGroupItem';
NotificationListGroupItem.defaultProps = {
  collapsed: false,
  showCounter: false,
  busy: false,
  priority: Priority.None,
  showClose: false,
  selected: false
};

export { NotificationListGroupItem };
