// @ui5/webcomponents-react/lib/NotificationListItem.js
import { Priority } from '@ui5/webcomponents-react/lib/Priority';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents-fiori/dist/NotificationListItem';

/**
 * The `NotificationListItem` is a type of list item, meant to display notifications.
 * The component has a rich set of various properties that allows the user to set `avatar`, `heading`, descriptive `content` and `footnotes` to fully describe a notification.
 * The user can:
 *
 * *   display a `Close` button
 * *   can control whether the `heading` and `description` should wrap or truncate and display a `ShowMore` button to switch between less and more information
 * *   add custom actions by using the `NotificationAction` component
 *
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/NotificationListItem" target="_blank">UI5 Web Components Playground</a>
 */
var NotificationListItem = withWebComponent('ui5-li-notification', ['heading', 'priority'], ['read', 'wrap', 'busy', 'showClose', 'selected'], ['avatar', 'footnotes', 'actions'], ['close']);
NotificationListItem.displayName = 'NotificationListItem';
NotificationListItem.defaultProps = {
  read: false,
  wrap: false,
  busy: false,
  priority: Priority.None,
  showClose: false,
  selected: false
};

export { NotificationListItem };
