// @ui5/webcomponents-react/lib/RadioButton.js
import { ValueState } from '@ui5/webcomponents-react/lib/ValueState';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/RadioButton';

/**
 * The `RadioButton` component enables users to select a single option from a set of options. When a `RadioButton` is selected by the user, the `onSelect` event is fired. When a `RadioButton` that is within a group is selected, the one that was previously selected gets automatically deselected. You can group radio buttons by using the `name` property.
 * **Note:** If `RadioButton` is not part of a group, it can be selected once, but can not be deselected back.
 *
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/RadioButton" target="_blank">UI5 Web Components Playground</a>
 */
var RadioButton = withWebComponent('ui5-radiobutton', ['name', 'text', 'value', 'valueState'], ['disabled', 'readonly', 'selected', 'wrap'], [], ['select']);
RadioButton.displayName = 'RadioButton';
RadioButton.defaultProps = {
  disabled: false,
  readonly: false,
  selected: false,
  valueState: ValueState.None,
  wrap: false
};

export { RadioButton };
