// @ui5/webcomponents-react/lib/TabContainer.js
import { TabContainerTabsPlacement } from '@ui5/webcomponents-react/lib/TabContainerTabsPlacement';
import { TabLayout } from '@ui5/webcomponents-react/lib/TabLayout';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/TabContainer';

/**
 * The `TabContainer` represents a collection of tabs with associated content. Navigation through the tabs changes the content display of the currently active content area. A tab can be labeled with text only, or icons with text.
 *
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/TabContainer" target="_blank">UI5 Web Components Playground</a>
 */
var TabContainer = withWebComponent('ui5-tabcontainer', ['tabLayout', 'tabsPlacement'], ['collapsed', 'fixed', 'showOverflow'], ['overflowButton'], ['tab-select']);
TabContainer.displayName = 'TabContainer';
TabContainer.defaultProps = {
  collapsed: false,
  fixed: false,
  showOverflow: false,
  tabLayout: TabLayout.Standard,
  tabsPlacement: TabContainerTabsPlacement.Top
};

export { TabContainer };
