// @ui5/webcomponents-react/lib/TextArea.js
import { ValueState } from '@ui5/webcomponents-react/lib/ValueState';
import { withWebComponent } from '@ui5/webcomponents-react/lib/withWebComponent';
import '@ui5/webcomponents/dist/TextArea';

/**
 * The `TextArea` component provides large spaces for text entries in the form of multiple rows. It has the functionality of the `TextField` with the additional functionality for multiline texts.
 *
 * When empty, it can hold a placeholder similar to a `Input`. You can define the rows of the `TextArea` and also determine specific behavior when handling long texts.
 *
 * <a href="https://sap.github.io/ui5-webcomponents/playground/components/TextArea" target="_blank">UI5 Web Components Playground</a>
 */
var TextArea = withWebComponent('ui5-textarea', ['growingMaxLines', 'maxlength', 'name', 'placeholder', 'rows', 'value', 'valueState'], ['disabled', 'growing', 'readonly', 'required', 'showExceededText'], ['valueStateMessage'], ['change', 'input']);
TextArea.displayName = 'TextArea';
TextArea.defaultProps = {
  disabled: false,
  growing: false,
  growingMaxLines: 0,
  maxlength: null,
  readonly: false,
  required: false,
  rows: 0,
  showExceededText: false,
  valueState: ValueState.None
};

export { TextArea };
