// @ui5/webcomponents-react/lib/Text.js
import _extends from '@babel/runtime/helpers/extends';
import { createUseStyles } from 'react-jss';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { usePassThroughHtmlProps } from '@ui5/webcomponents-react-base/lib/usePassThroughHtmlProps';
import React, { forwardRef } from 'react';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';

var TextStyles = {
  // Text Style
  text: {
    fontFamily: ThemingParameters.sapFontFamily,
    fontSize: ThemingParameters.sapFontSize,
    fontWeight: 'normal',
    color: ThemingParameters.sapTextColor,
    display: 'inline-block',
    boxSizing: 'border-box',
    whiteSpace: 'pre-line',
    wordWrap: 'break-word',
    maxWidth: '100%'
  },
  renderWhitespace: {
    whiteSpace: 'pre-wrap'
  },
  noWrap: {
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    wordWrap: 'normal',
    overflow: 'hidden',
    '&$renderWhitespace': {
      whiteSpace: 'pre'
    }
  }
};

var useStyles = createUseStyles(TextStyles, {
  name: 'Text'
});
/**
 * The `Text` component can be used for embedding text into your app. You can hyphenate the text with the use of the `wrapping` prop.
 * <br />__Note:__ Line breaks will always be visualized except when the wrapping property is set to false. In addition, tabs and whitespace can be preserved by setting the renderWhitespace property to true.
 */

var Text = /*#__PURE__*/forwardRef(function (props, ref) {
  var children = props.children,
      renderWhitespace = props.renderWhitespace,
      wrapping = props.wrapping,
      className = props.className,
      style = props.style,
      tooltip = props.tooltip,
      slot = props.slot;
  var classes = useStyles();
  var classNameString = StyleClassHelper.of(classes.text);

  if (wrapping === false) {
    classNameString.put(classes.noWrap);
  }

  if (renderWhitespace) {
    classNameString.put(classes.renderWhitespace);
  }

  if (className) {
    classNameString.put(className);
  }

  var passThroughProps = usePassThroughHtmlProps(props);
  return /*#__PURE__*/React.createElement("span", _extends({
    ref: ref,
    style: style,
    className: classNameString.toString(),
    title: tooltip,
    slot: slot
  }, passThroughProps), children);
});
Text.defaultProps = {
  renderWhitespace: false,
  wrapping: true
};
Text.displayName = 'Text';

export { Text };
