// @ui5/webcomponents-react/lib/ToolbarSeparator.js
import { useI18nBundle } from '@ui5/webcomponents-react-base/lib/hooks';
import { createUseStyles } from 'react-jss';
import { CssSizeVariables } from '@ui5/webcomponents-react-base/lib/CssSizeVariables';
import { StyleClassHelper } from '@ui5/webcomponents-react-base/lib/StyleClassHelper';
import { ThemingParameters } from '@ui5/webcomponents-react-base/lib/ThemingParameters';
import React from 'react';
import { SEPARATOR } from '@ui5/webcomponents-react/dist/assets/i18n/i18n-defaults';

var styles = {
  separator: {
    width: '0.0625rem',
    height: CssSizeVariables.sapWcrToolbarSeparatorHeight,
    background: ThemingParameters.sapToolbar_SeparatorColor
  }
};
var useStyles = createUseStyles(styles, {
  name: 'ToolbarSeparator'
});

var ToolbarSeparator = function ToolbarSeparator(props) {
  var style = props.style,
      className = props.className;
  var classes = useStyles(styles);
  var separatorClasses = StyleClassHelper.of(classes.separator);
  var i18nBundle = useI18nBundle('@ui5/webcomponents-react');

  if (className) {
    separatorClasses.put(className);
  }

  return /*#__PURE__*/React.createElement("div", {
    style: style,
    className: separatorClasses.valueOf(),
    "aria-label": i18nBundle.getText(SEPARATOR)
  });
};

ToolbarSeparator.displayName = 'ToolbarSeparator';

export { ToolbarSeparator };
